src/algebra/big_operators/nat_antidiagonal.lean

```diff
@@ -54,5 +54,19 @@ lemma prod_antidiagonal_subst {n : ℕ} {f : ℕ × ℕ → ℕ → M} :
   ∏ p in antidiagonal n, f p n = ∏ p in antidiagonal n, f p (p.1 + p.2) :=
 prod_congr rfl $ λ p hp, by rw [nat.mem_antidiagonal.1 hp]
 
+@[to_additive]
+lemma prod_antidiagonal_eq_prod_range_succ {M : Type*} [comm_monoid M] (f : ℕ → ℕ → M) (n : ℕ) :
+  ∏ ij in finset.nat.antidiagonal n, f ij.1 ij.2 = ∏ k in range n.succ, f k (n - k) :=
+begin
+  refine finset.prod_bij' (λ (ij : ℕ × ℕ) _, ij.1) _ _
+    (λ a _, (a, n - a) : Π (a : ℕ), a ∈ finset.range n.succ → ℕ × ℕ) _ _ _,
+  { rintro ⟨i, j⟩ ha, rw mem_antidiagonal at ha, simp [mem_range_succ_iff, nat.le.intro ha] },
+  { rintro ⟨i, j⟩ ha, rw mem_antidiagonal at ha, subst ha, congr', simp },
+  { intros a ha, rw mem_range_succ_iff at ha, rw mem_antidiagonal, exact nat.add_sub_of_le ha, },
+  { rintro ⟨i, j⟩ ha, rw mem_antidiagonal at ha, subst ha, congr', simp },
+  { rintros, refl },
+end
```

Comment:
This can be golfed to:
```suggestion
@[to_additive]
lemma prod_antidiagonal_eq_prod_range_succ {M : Type*} [comm_monoid M] (f : ℕ → ℕ → M) (n : ℕ) :
  ∏ ij in finset.nat.antidiagonal n, f ij.1 ij.2 = ∏ k in range n.succ, f k (n - k) :=
calc ∏ ij in finset.nat.antidiagonal n, f ij.1 ij.2
    = ∏ ij in (range n.succ).map ⟨λ i, (i, n - i), λ x y h, (prod.mk.inj h).1⟩, f ij.1 ij.2 :
              by refl
... = ∏ k in range n.succ, f k (n - k) : by rw [prod_map, function.embedding.coe_fn_mk]
```

Reply:
Or with less repetition:
```suggestion
@[to_additive]
lemma prod_antidiagonal_eq_prod_range_succ {M : Type*} [comm_monoid M] (f : ℕ → ℕ → M) (n : ℕ) :
  ∏ ij in finset.nat.antidiagonal n, f ij.1 ij.2 = ∏ k in range n.succ, f k (n - k) :=
begin
  change ∏ ij in (range n.succ).map ⟨λ i, (i, n - i), λ x y h, (prod.mk.inj h).1⟩, f ij.1 ij.2 = _,
  rw [prod_map, function.embedding.coe_fn_mk]
end
```

Reply:
I golfed it even more. Thanks!